xds/internal/client/client.go

```diff
@@ -129,14 +130,52 @@ type loadReportingOptions struct {
 // resource updates from an APIClient for a specific version.
 type UpdateHandler interface {
 	// NewListeners handles updates to xDS listener resources.
-	NewListeners(map[string]ListenerUpdate)
+	NewListeners(map[string]ListenerUpdate, UpdateMetadata)
 	// NewRouteConfigs handles updates to xDS RouteConfiguration resources.
-	NewRouteConfigs(map[string]RouteConfigUpdate)
+	NewRouteConfigs(map[string]RouteConfigUpdate, UpdateMetadata)
 	// NewClusters handles updates to xDS Cluster resources.
-	NewClusters(map[string]ClusterUpdate)
+	NewClusters(map[string]ClusterUpdate, UpdateMetadata)
 	// NewEndpoints handles updates to xDS ClusterLoadAssignment (or tersely
 	// referred to as Endpoints) resources.
-	NewEndpoints(map[string]EndpointsUpdate)
+	NewEndpoints(map[string]EndpointsUpdate, UpdateMetadata)
+}
+
+// ServiceStatus is the status of the update.
+type ServiceStatus int
+
+const (
+	ServiceStatusUnknown ServiceStatus = iota
+	ServiceStatusRequested
+	ServiceStatusNotExist // Resource is removed in the server, in LDS/CDS.
+	ServiceStatusACKed
+	ServiceStatusNACKed
+)
+
+// UpdateErrorMetadata is part of UpdateMetadata. It contains the error state
+// when a response is NACKed.
+type UpdateErrorMetadata struct {
+	// Version is the version of the NACKed response.
+	Version string
+	// Err contains why the response was NACKed.
+	Err error
+	// Timestamp is when the NACKed response was received.
+	Timestamp time.Time
+}
+
+// UpdateMetadata contains the metadata for each update, including timestamp,
+// raw message, and so on.
+type UpdateMetadata struct {
```

Comment:
Did you consider structuring the data structures as follows:

```golang
type UpdateMetadata struct {
        Raw        *anypb.Any // Last good response 
        Status     ServiceStatus
        AckVersion string
        Timestamp  time.Time
}

type UpdateErrorMetadata struct {
        Raw         *anypb.Any // Last error-ed response
        NackVersion string
        Err         error
        Timestamp   time.Time
}

type ListenerUpdate struct {
        Parsed ParsedListener
        MD     *UpdateMetadata
        ErrMD  *UpdateErrorMetadata
}

type ParsedListener struct {
        // This contains all the fields of the existing ListenerUpdate struct.
}
```

With this approach, the `NewXxx()` methods like `NewListener()` need not change in its signature.

This also reduces the number of fields in the `clientImpl` per resource type.
```golang
type clientImpl struct {
  ...
  ldsWatchers map[string]map[*watchInfo]bool
  ldsCache    map[string]ListenerUpdate // This contains everything
  ...
}
```

Reply:
If there is a concern that `ListenerUpdate` now has fields which are not required by callers of the `WatchListener`, we can make those fields unexported. And since the CSDS servicer implementation will probably live in the same `client` package, it will have access to those fields.

Reply:
This is similar to what I started with, but I didn't like the idea later.
`Update` is per resource, `Metadata` is per xDS message. They are at different levels.
So `NewListeners(updates map[string]ListenerUpdate, metadata UpdateMetadata)` takes a list of `Update`s, but just one `Metadata`.

Merging `Metadata` into `Update` makes it look like in the same `NewListener()` call, there can be multiple `Metadata`.
(It's unfortunate that the cache eventually needs to make copies of `Metadata` to keep different versions)